### DISCUSSION TITLE
Git workflow questions

### DISCUSSION THREAD

**Discussion by dmastrop** [community/community#51071] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Product Feedback

### Body

hi all,

I have a question about how to synch up local git repos and remote git repos with 2 branches (for example), after making a change in one of the branches. I was able to get this to work but my method seems to be overly complex.

The setup is a workflow where any commit to the master branch will kick off a git workflow script and deploy the code to AWS (Elasticbeanstalk as a docker container).

I have a feature branch and this master branch.

I made a local change to the feature branch 
**git checkout -b feature** (I created the branch and checked it out for this first time)
I made a change to the branch (one of the .js files)
**git add .
git commit -m "changes to js file"
git push origin feature**

So I see the changes pushed to the remote feature branch git repo
I next wanted to synch this up with the remote master branch to kick off a deployment to the AWS

I clicked the Compare and Pull Request

Clicked Create pull request

I then did a Merge pull request (merged feature to master branch)

This kicked off the scripts, the code was tested and deployed to AWS.

When I went to the local master git repo I get this error below when trying to do a synch to the remote master repo

**git checkout master
git push origin master**
To https://github.com/xxxxxx/docker-react.git
 ! [rejected]        master -> master (fetch first)
**error: failed to push some refs to 'https://github.com/dmastrop/docker-react.git'
hint: Updates were rejected because the remote contains work that you do
hint: not have locally. This is usually caused by another repository pushing
hint: to the same ref.** **You may want to first integrate the remote changes
hint: (e.g., 'git pull ...') before pushing again.**
hint: See the 'Note about fast-forwards' in 'git push --help' for details.

**_I know why this happened (due to the remote merge of feature to master), and was curious as to why, after merging feature to master on remote, that the master remote does not push changes to the local master repo._**

To get everything in order again I did what was recommended and did a manual pull on the local from the remote master

**git checkout master**
**git pull origin master**

and this synched up the local master repo with the remote master repo so that I could start making changes  on master again and pushing to master remote.

_**Is this the way to do things when committing to a feature branch and then to master, to get everything synched up between local and remote repos?  My approach seems a bit messy**_

_**Also could I have alternatively used the git merge on the local repo side to get the master in synch with the feature branch?**_

**git checkout master
git merge feature**

**_Does the git merge feature require the username i.e,_**
**git merge (username)/feature**     ?

warm regards

**Comment by airtower-luna** [comment, marked as answer]
> I know why this happened (due to the remote merge of feature to master), and was curious as to why, after merging feature to master on remote, that the master remote does not push changes to the local master repo.

Obvious reason: GitHub doesn't have access to your local machine to push to it.

In general, Git doesn't do any remote access unless you explicitly tell it to. Doing a pull before working on a branch that has been updated upstream is exactly the right approach. If you forget and already have made commits and then can't push, you can rebase on top of the remote branch.

**Reply by dmastrop** [reply]
Thanks AirTower
Sorry for basic question, but what do you mean by "rebase on top of the remote branch".  I understand your point and luckily I did not make any commits prior to doing that first "git push origin master" that failed.
I assume this rebase would somehow get my new local commits + the remote master all merged together in that master branch.

warm regards

**Reply by airtower-luna** [reply]
The very short version is, rebasing recreates your commits as if you had started work from the branch you're rebasing onto. Which can be a remote branch that you have fetched, e.g. `git rebase origin/main` will rebase onto the `main` branch in the `origin` remote.

Long version: See the [Git Branching - Rebasing](https://git-scm.com/book/en/v2/Git-Branching-Rebasing) chapter in the Git Book (highly recommended in general).

**Reply by dmastrop** [reply]
thanks
